src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -984,6 +985,10 @@ trait Contexts { self: Analyzer =>
 
     def isNameInScope(name: Name) = lookupSymbol(name, _ => true).isSuccess
 
+    def isPlainClassInJavaUnit(sym: Symbol): Boolean =
```

Comment:
I agree with your point of moving the `sym.isJavaDefined` check into `FindMember`. However, your example is not going to work, and it will be very hard to support it:
```
// A.java
public class A {
  public static class C {
    public static class I { }
  }
  public static class E extends D.I { }
}
```

```
// Test.scala
class D extends A.C

object Test {
  new A.E()
}
```

In mixed compilation:

```
A.java:5: error: not found: value D
  public static class E extends D.I { }
                                ^
one error found
```